src/nccl_ofi_ofiutils.c

```diff
@@ -259,7 +259,12 @@ int nccl_ofi_ofiutils_init_connection(int api_version, struct fi_info *info, str
 		goto error;
 	}
 
-	cq_attr.format = FI_CQ_FORMAT_TAGGED;
+	if (info->caps & FI_TAGGED) {
+		cq_attr.format = FI_CQ_FORMAT_TAGGED;
+	} else {
+		cq_attr.format = FI_CQ_FORMAT_DATA;
+	}
```

Comment:
I don't think this is a valid approach. From the Libfabric doc:
> Providers may indicate support for additional capabilities beyond those requested when the use of expanded capabilities will not adversely affect performance or expose the application to communication beyond that which was requested. Applications may use this feature to request a minimal set of requirements, then check the returned capabilities to enable additional optimizations.

So, even though for RDMA protocol we don't request `FI_TAGGED` capability, the provider may still return it. If that happens, `nccl_ofi_ofiutils_init_connection` will set `cq_attr.format == FI_CQ_FORMAT_TAGGED`, but the code in `nccl_ofi_rdma.c` assumes the format is `FI_CQ_FORMAT_DATA`: https://github.com/aws/aws-ofi-nccl/blob/e3f9fd4b87fb18111c6ee43f4e6ac041f7803fb5/src/nccl_ofi_rdma.c#L1674

I think we may need to go back to adding `tagged` parameter to `nccl_ofi_ofiutils_init_connection`

Reply:
`FI_TAGGED` is a primary capability. Providers must enable it only when explicitly requested. The part you cite applies to secondary capabilities (which is the loose wording around _not adversely affect performance_, which tag matching would)

From fi_getinfo(3):

> Capabilities may be grouped into three general categories: primary, secondary, and primary modifiers. Primary capabilities must explicitly be requested by an application, and a provider must enable support for only those primary capabilities which were selected. Primary modifiers are used to limit a primary capability, such as restricting an endpoint to being send-only. If no modifiers are specified for an applicable capability, all relevant modifiers are assumed. See above definitions for details.

Reply:
To support Raghu's point, I did test this before pushing it and confirmed that Libfabric was setting the bit only if the capability was requested.